File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        def isSameTree(p, q):
            if not p and not q:
                return True
            elif not p or not q:
                return False
            else:
                if p.val != q.val:
                    return False
                
                return isSameTree(p.left, q.left) and isSameTree(p.right, q.right) 

        if not root:
            return False

        queue = deque([root])

        while queue:
            node = queue.popleft()

            if node.val == subRoot.val and isSameTree(node, subRoot):
                return True
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        return False


        

File: Data Structures & Algorithms/invert-a-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {TreeNode}
     */
    

    invertTree(root) {
        if (!root) return root

        if (root.left) {
            this.invertTree(root.left)
        }
        if (root.right) {
            this.invertTree(root.right)
        }
        [root.left, root.right] = [root.right, root.left]

        return root
    }
}
